src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1238 @@
+ *
+ * Per try we increase the delay as follows:
+ *   retry 1: 1 min
+ *   retry 2: 10 min
+ *   retry 3: 60 min
+ *
+ * In the future we would like a callback on the job_type that could
+ * distinguish the retry count and delay + potential jitter on a
+ * job_type basis. For now we only assume this to be used by the
+ * rebalancer and settled on the retry scheme above.
+ */
+static int64
+CalculateBackoffDelay(int retryCount)
+{
+	switch (retryCount)
+	{
+		case 1:
+		{
+			return 1 * 60 * 1000;
+		}
+
+		case 2:
+		{
+			return 10 * 60 * 1000;
+		}
+
+		case 3:
+		default: /* just make sure retrying an hour for missed cases */
+		{
+			return 60 * 60 * 1000;
+		}
+	}
+}
+
+
+#if PG_VERSION_NUM < PG_VERSION_15
+static const char *
+error_severity(int elevel)
+{
+	const char *prefix;
+
+	switch (elevel)
+	{
+		case DEBUG1:
+		case DEBUG2:
+		case DEBUG3:
+		case DEBUG4:
+		case DEBUG5:
+		{
+			prefix = gettext_noop("DEBUG");
+			break;
+		}
+
+		case LOG:
+		case LOG_SERVER_ONLY:
+		{
+			prefix = gettext_noop("LOG");
+			break;
+		}
+
+		case INFO:
+		{
+			prefix = gettext_noop("INFO");
+			break;
+		}
+
+		case NOTICE:
+		{
+			prefix = gettext_noop("NOTICE");
+			break;
+		}
+
+		case WARNING:
+		{
+			prefix = gettext_noop("WARNING");
+			break;
+		}
+
+#if PG_VERSION_NUM >= PG_VERSION_14
+		case WARNING_CLIENT_ONLY:
+		{
+			prefix = gettext_noop("WARNING");
+			break;
+		}
+#endif
+
+		case ERROR:
+		{
+			prefix = gettext_noop("ERROR");
+			break;
+		}
+
+		case FATAL:
+		{
+			prefix = gettext_noop("FATAL");
+			break;
+		}
+
+		case PANIC:
+		{
+			prefix = gettext_noop("PANIC");
+			break;
+		}
+
+		default:
+		{
+			prefix = "???";
+			break;
+		}
+	}
+
+	return prefix;
+}
+
+
+#endif
+
+
+/*
+ * bgw_generate_returned_message -
+ *      generates the message to be inserted into the job_run_details table
+ *      first part is comming from error_severity (elog.c)
+ */
+static void
+bgw_generate_returned_message(StringInfoData *display_msg, ErrorData edata)
+{
+	const char *prefix = error_severity(edata.elevel);
+	appendStringInfo(display_msg, "%s: %s", prefix, edata.message);
+	if (edata.detail != NULL)
+	{
+		appendStringInfo(display_msg, "\nDETAIL: %s", edata.detail);
+	}
+
+	if (edata.hint != NULL)
+	{
+		appendStringInfo(display_msg, "\nHINT: %s", edata.hint);
+	}
+
+	if (edata.context != NULL)
+	{
+		appendStringInfo(display_msg, "\nCONTEXT: %s", edata.context);
+	}
+}
+
+
+/*
+ * UpdateDependingTasks updates all depending tasks, based on the type of termnial state
+ * the current task reached.
+ */
+static void
+UpdateDependingTasks(BackgroundTask *task)
+{
+	switch (task->status)
+	{
+		case BACKGROUND_TASK_STATUS_DONE:
+		{
+			UnblockDependingBackgroundTasks(task->jobid, task->taskid);
+			break;
+		}
+
+		case BACKGROUND_TASK_STATUS_ERROR:
+		{
+			/* when we error this task, we need to unschedule all dependant tasks */
+			UnscheduleDependantTasks(task->jobid, task->taskid);
```

Comment:
maybe it's nicer to pass the task to these functions directly, instead of extracting jobid+taskid from it.

Reply:
makes sense, I think there was a moment in time where one of the functions was called separately from a place where the struct wasn't available. Can't seem to find that now.